internal/lsp/lspservertests/testserver.go

```diff
@@ -0,0 +1,254 @@
+package lspservertests
+
+import (
+	"fmt"
+	"strings"
+	"testing"
+
+	"github.com/go-json-experiment/json"
+	"github.com/go-json-experiment/json/jsontext"
+	"github.com/microsoft/typescript-go/internal/bundled"
+	"github.com/microsoft/typescript-go/internal/collections"
+	"github.com/microsoft/typescript-go/internal/ls/lsconv"
+	"github.com/microsoft/typescript-go/internal/lsp/lsproto"
+	"github.com/microsoft/typescript-go/internal/project"
+	"github.com/microsoft/typescript-go/internal/testutil/fsbaselineutil"
+	"github.com/microsoft/typescript-go/internal/testutil/lsptestutil"
+	"github.com/microsoft/typescript-go/internal/testutil/projecttestutil"
+	"gotest.tools/v3/assert"
+)
+
+type testServer struct {
+	t             *testing.T
+	files         map[string]any
+	server        *lsptestutil.TestLspServer
+	utils         *projecttestutil.SessionUtils
+	baseline      strings.Builder
+	fsDiffer      *fsbaselineutil.FSDiffer
+	writtenFiles  collections.SyncSet[string]
+	isInitialized bool
+
+	serializedProjects           map[string]projectInfo
+	serializedOpenFiles          map[string]*openFileInfo
+	serializedConfigFileRegistry *project.ConfigFileRegistry
+	openFiles                    map[string]string
+}
+
+func newTestServer(t *testing.T, files map[string]any) *testServer {
+	t.Helper()
+	server, utils := lsptestutil.Setup(t, files)
+	testServer := &testServer{
+		t:         t,
+		files:     files,
+		server:    server,
+		utils:     utils,
+		openFiles: make(map[string]string),
+	}
+	testServer.fsDiffer = &fsbaselineutil.FSDiffer{
+		FS:           utils.FsFromFileMap(),
+		WrittenFiles: &testServer.writtenFiles,
+	}
+	fmt.Fprintf(&testServer.baseline, "UseCaseSensitiveFileNames: %v\n", utils.FsFromFileMap().UseCaseSensitiveFileNames())
+	testServer.fsDiffer.BaselineFSwithDiff(&testServer.baseline)
+	return testServer
+}
+
+func (s *testServer) content(fileName string) string {
+	if text, ok := s.openFiles[fileName]; ok {
+		return text
+	}
+	return s.files[fileName].(string)
+}
+
+func (s *testServer) hoverToWriteProjectStatus(fileName string) {
+	// Do hover so we have snapshot to check things on!!
+	_, _, resultOk := lsptestutil.SendRequest(s.t, s.server, lsproto.TextDocumentHoverInfo, &lsproto.HoverParams{
+		TextDocument: lsproto.TextDocumentIdentifier{
+			Uri: lsproto.DocumentUri("file://" + fileName),
```

Comment:
Maybe this is a pure code move from somewhere else, but this is technically wrong given escaping and such means we should be using the func that converts these.